--- a/source/blender/blenlib/BLI_linear_allocator.hh
+++ b/source/blender/blenlib/BLI_linear_allocator.hh
@@ -36,7 +36,7 @@ template<typename Allocator = GuardedAllocator> class LinearAllocator : NonCopya
 
   /* Buffers larger than that are not packed together with smaller allocations to avoid wasting
    * memory. */
-  constexpr static int64_t large_buffer_threshold = 4096;
+  constexpr static int64_t large_buffer_threshold = 0;
 
  public:
 #ifdef BLI_DEBUG_LINEAR_ALLOCATOR_SIZE
